### ISSUE TITLE
Moq1001/Moq1002 false negative: target-typed new (Mock<IFoo> m = new(42);) is never analyzed

### ISSUE THREAD

**Issue by rjmurillo** [rjmurillo/moq.analyzers#1245]
## Summary

An internal code audit found that `ConstructorArgumentsShouldMatchAnalyzer` registers its object-creation callback only for `SyntaxKind.ObjectCreationExpression`. Target-typed `new` (C# 9) produces a distinct node kind, `SyntaxKind.ImplicitObjectCreationExpression`, so `Mock<IFoo> m = new(42);` and `Mock<Foo> f = new(1, true);` are silently skipped — constructor-argument mismatches that fail at runtime with a `MockException` get no **Moq1001**/**Moq1002** diagnostic. **Severity: Medium (false negative on an increasingly common syntax form).** The recommended fix is the incremental one: also register `ImplicitObjectCreationExpression` and rework the handler around `BaseObjectCreationExpressionSyntax`; a full IOperation migration (ADR-007) stays out of scope.

## Current behavior

Registration — `src/Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs:164-166`:

```csharp
        // These are for classes
        context.RegisterSyntaxNodeAction(context => AnalyzeNewObject(context, knownSymbols), SyntaxKind.ObjectCreationExpression);
        context.RegisterSyntaxNodeAction(context => AnalyzeInstanceCall(context, knownSymbols), SyntaxKind.InvocationExpression);
```

Handler — `src/Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs:210-244` (hard cast at line 218):

```csharp
    /// <summary>
    /// Analyzes when a Mock`1 object is created to verify the provided constructor arguments
    /// match an existing constructor of the mocked class.
    /// </summary>
    /// <param name="context">The context.</param>
    /// <param name="knownSymbols">The <see cref="MoqKnownSymbols"/> used to lookup symbols against Moq types.</param>
    private static void AnalyzeNewObject(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        ObjectCreationExpressionSyntax newExpression = (ObjectCreationExpressionSyntax)context.Node;

        GenericNameSyntax? genericNameSyntax = GetGenericNameSyntax(newExpression.Type);
        if (genericNameSyntax == null)
        {
            return;
        }

        SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo(newExpression, context.CancellationToken);

        if (!symbolInfo
            .Symbol?
            .IsInstanceOf(knownSymbols.Mock1?.Constructors ?? ImmutableArray<IMethodSymbol>.Empty)
            ?? false)
        {
            return;
        }

        if (symbolInfo.Symbol?.ContainingType is not INamedTypeSymbol { IsGenericType: true } typeSymbol)
        {
            return;
        }

        ITypeSymbol mockedClass = typeSymbol.TypeArguments[0];

        VerifyMockAttempt(context, knownSymbols, mockedClass, newExpression.ArgumentList, true);
    }
```

**Failure scenario** (no diagnostic today; explicit-`new` spelling of the same code is flagged):

```csharp
public interface IFoo { }

Mock<IFoo> m = new(42);          // expected Moq1001 on `(42)` — actual: nothing
Mock<Foo>  f = new("1", 3);      // expected Moq1002 on `("1", 3)` — actual: nothing
```

**Where the mocked type `T` comes from (the key design point, verified).** The current handler does *not* extract `T` from the syntax. `GetGenericNameSyntax(newExpression.Type)` (lines 52-67) is only a cheap **syntactic pre-filter**; the mocked type is derived semantically from the bound constructor: `symbolInfo.Symbol.ContainingType` is `Moq.Mock<T>` and `typeSymbol.TypeArguments[0]` is `T`. For an implicit object creation there is **no type syntax at all**, but `SemanticModel.GetSymbolInfo` on an `ImplicitObjectCreationExpressionSyntax` returns the same bound constructor symbol as the explicit form (the target type flows from the declaration/assignment), so the entire semantic pipeline (`IsInstanceOf(Mock1.Constructors)` → `ContainingType.TypeArguments[0]` → `VerifyMockAttempt`) works unchanged. The only required changes are: (a) register the extra syntax kind, (b) type the local as `BaseObjectCreationExpressionSyntax` (the common base, already used elsewhere in the repo — see `src/CodeFixes/CallbackSignatureShouldMatchMockedMethodFixer.cs:151`), and (c) apply the syntactic pre-filter only to the explicit form. If the target type is unresolvable (mid-edit), `symbolInfo.Symbol` is `null` and the handler bails, consistent with the `IErrorTypeSymbol` guard in `VerifyMockAttempt` (line 414). No `GetTypeInfo` call is needed.

**Diagnostic location (verified).** All report paths use the argument list span: `ConstructorIsEmpty` (`src/Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs:398-405`) computes `argumentList?.GetLocation() ?? context.Node.GetLocation()`, and `VerifyInterfaceMockAttempt`/`VerifyDelegateMockAttempt` create the diagnostic on `argumentList`. `ImplicitObjectCreationExpressionSyntax` grammatically always has an argument list (`new(...)` — the parentheses are mandatory), so implicit-new diagnostics land on `(42)`, exactly like the explicit form.

## Desired behavior

`Mock<T> m = new(args);` (local/field/property initializers, assignments, return statements — any target-typed context) is analyzed identically to `new Mock<T>(args)`: Moq1001 for interface mocks with constructor arguments, Moq1002 for classes with no matching constructor, with the diagnostic on the argument-list span.

## Implementation plan

1. **Registration** — in `src/Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs`, replace line 165 with (the `RegisterSyntaxNodeAction` overload takes `params TLanguageKindEnum[]`, so one registration covers both kinds):

```csharp
        context.RegisterSyntaxNodeAction(context => AnalyzeNewObject(context, knownSymbols), SyntaxKind.ObjectCreationExpression, SyntaxKind.ImplicitObjectCreationExpression);
```

2. **Handler** — replace the whole `AnalyzeNewObject` method (lines 210-244, including its XML doc) with:

```csharp
    /// <summary>
    /// Analyzes when a Mock`1 object is created to verify the provided constructor arguments
    /// match an existing constructor of the mocked class. Handles both explicit
    /// (<c>new Mock&lt;T&gt;(...)</c>) and target-typed (<c>Mock&lt;T&gt; m = new(...);</c>) creation.
    /// </summary>
    /// <param name="context">The context.</param>
    /// <param name="knownSymbols">The <see cref="MoqKnownSymbols"/> used to lookup symbols against Moq types.</param>
    private static void AnalyzeNewObject(SyntaxNodeAnalysisContext context, MoqKnownSymbols knownSymbols)
    {
        BaseObjectCreationExpressionSyntax newExpression = (BaseObjectCreationExpressionSyntax)context.Node;

        // Cheap syntactic pre-filter for the explicit form: `new Mock<T>(...)` must name a
        // generic type. The implicit form (`Mock<T> m = new(...);`) has no type syntax, so
        // it goes straight to symbol resolution; the mocked type is derived from the bound
        // constructor symbol below, not from the syntax.
        if (newExpression is ObjectCreationExpressionSyntax explicitNew
            && GetGenericNameSyntax(explicitNew.Type) is null)
        {
            return;
        }

        SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo(newExpression, context.CancellationToken);

        if (!symbolInfo
            .Symbol?
            .IsInstanceOf(knownSymbols.Mock1?.Constructors ?? ImmutableArray<IMethodSymbol>.Empty)
            ?? false)
        {
            return;
        }

        if (symbolInfo.Symbol?.ContainingType is not INamedTypeSymbol { IsGenericType: true } typeSymbol)
        {
            return;
        }

        ITypeSymbol mockedClass = typeSymbol.TypeArguments[0];

        VerifyMockAttempt(context, knownSymbols, mockedClass, newExpression.ArgumentList, true);
    }
```

   Notes:
   - `BaseObjectCreationExpressionSyntax.ArgumentList` is the shared nullable property; for implicit creations it is always non-null, for explicit creations with only an initializer (`new Mock<T> { }`) it is null exactly as today — `VerifyMockAttempt` already handles a null `argumentList`.
   - `ImplicitObjectCreationExpressionSyntax`/`SyntaxKind.ImplicitObjectCreationExpression` exist since Roslyn 3.8, well within the pinned 4.8 API (ADR-003).
   - No using changes; every type referenced is already used in this file or its global usings.
   - Perf: implicit creations get one `GetSymbolInfo` call each without a syntactic pre-filter. This matches the analyzer's existing cost profile (it already semantically binds **every** invocation at line 181), and there is no cheap syntactic discriminator for `new(...)`. Broader binding reduction is the ADR-007 migration, out of scope.
   - Keep `GetGenericNameSyntax` (lines 52-67) unchanged — still used for the explicit-form pre-filter.

3. **Docs** — behavior change (new syntax form now covered): add a target-typed `new` example to `docs/rules/Moq1001.md` and `docs/rules/Moq1002.md` (e.g. `Mock<IMyService> mock = new("123"); // Moq1001` next to the existing explicit-new examples). Do **not** touch `AnalyzerReleases.Shipped.md`; no `AnalyzerReleases.Unshipped.md` entry (no severity/category change).

## Tests

Test-infra language version (verified): the test compiler is the pinned Microsoft.CodeAnalysis.CSharp **4.8** with `LanguageVersion.Default`, which is C# 12 for that compiler — target-typed `new` (C# 9) compiles fine; no parse-options changes needed.

**1. Class rows** — `tests/Moq.Analyzers.Test/ConstructorArgumentsShouldMatchAnalyzerTests.cs`, add to `TestData()` (e.g. after the existing `new Mock<Foo>...` rows at lines 20-32; fixtures `Foo`, `ClassWithParams`, `ClassWithDefaultParamCtor`, `ClassWithRequiredParamCtor` already exist in the template):

```csharp
            // Target-typed new (C# 9): must be analyzed exactly like `new Mock<T>(...)`
            ["""Mock<Foo> mock = new(false, 0);"""],
            ["""Mock<Foo> mock = new(MockBehavior.Default, true, 1);"""],
            ["""Mock<Foo> mock = new{|Moq1002:(1, true)|};"""],
            ["""Mock<Foo> mock = new{|Moq1002:(MockBehavior.Default, 2, true)|};"""],
            ["""Mock<ClassWithParams> mock = new(DateTime.Now, DateTime.Now);"""],
            ["""Mock<ClassWithDefaultParamCtor> mock = new();"""],
            ["""Mock<ClassWithRequiredParamCtor> mock = new{|Moq1002:()|};"""],
```

   These rows are automatically multiplied across 2 namespaces × 2 Moq versions (4.8.2, 4.18.4) by `.WithNamespaces().WithMoqReferenceAssemblyGroups()`; `Mock<T>` constructor shapes are identical across both supported Moq versions, so no version gating is needed. (`var` cannot be used with target-typed `new`; the declared type is what makes the form target-typed.)

**2. Interface rows** — `tests/Moq.Analyzers.Test/ConstructorArgumentsShouldMatchAnalyzerTests.Interfaces.cs`, add to the `all` set in `InterfaceTestData()` (after line 20):

```csharp
            // Target-typed new (C# 9)
            ["""Mock<IFoo> mock = new();"""],
            ["""Mock<IFoo> mock = new(MockBehavior.Default);"""],
            ["""Mock<IFoo> mock = new{|Moq1001:(42)|};"""],
            ["""Mock<IFoo> mock = new{|Moq1001:(MockBehavior.Default, 42)|};"""],
```

**3. Edge coverage that must keep passing (no new code needed, verify by running):**
   - MockBehavior-skip logic on implicit new (covered by the `new(MockBehavior.Default, ...)` rows above — exercises `IsArgumentMockBehavior` through the new code path).
   - The Doppelganger suite (`ConstructorArgumentsShouldMatchAnalyzerTests.Doppelganger.cs`) — custom `Mock` classes remain excluded because the gate is `knownSymbols.Mock1?.Constructors` symbol identity, which the implicit path shares.
   - All existing explicit-`new`, `Mock.Of<T>`, and `MockRepository.Create<T>` rows across the nine `ConstructorArgumentsShouldMatchAnalyzerTests.*` partials — the explicit-form path is behavior-identical (same pre-filter, same semantic checks).

No new test files; the existing `[Theory]`/`[MemberData]` methods (`ShouldAnalyzeConstructorArguments`, `ShouldAnalyzeInterface`) consume the new rows.

## Constraints

- Analyzer assemblies target netstandard2.0; Roslyn SDK pinned to Microsoft.CodeAnalysis **4.8** (ADR-003). `ImplicitObjectCreationExpressionSyntax`, `SyntaxKind.ImplicitObjectCreationExpression`, and `BaseObjectCreationExpressionSyntax` all exist in 4.8; use nothing newer.
- No new package dependencies (ADR-004/ADR-005).
- Symbol-based detection per ADR-001 (the mocked type comes from the bound constructor symbol, not syntax); `MoqKnownSymbols` created once per compilation per ADR-006 (unchanged); the full IOperation migration recommended by ADR-007 is explicitly deferred (see Out of scope).
- Tests in `tests/Moq.Analyzers.Test/` following repo patterns: `{|Moq1001:...|}`/`{|Moq1002:...|}` markup asserting ID + exact span; `[Theory]` + `[MemberData]` with `public static IEnumerable<object[]>`; rows multiplied across 2 namespaces × 2 Moq versions via `TestDataExtensions`.
- Update `docs/rules/Moq1001.md` and `docs/rules/Moq1002.md` (rule behavior now covers a new syntax form). Do not touch `AnalyzerReleases.Shipped.md`; no `AnalyzerReleases.Unshipped.md` change.

## Validation

```bash
dotnet format --verify-no-changes   # or `dotnet format` before committing
dotnet build /p:PedanticMode=true   # must be 0 warnings, 0 errors
dotnet test --settings ./build/targets/tests/test.runsettings
```

## Acceptance criteria

- [ ] `AnalyzeNewObject` is registered for both `ObjectCreationExpression` and `ImplicitObjectCreationExpression` and casts to `BaseObjectCreationExpressionSyntax`.
- [ ] `Mock<IFoo> m = new(42);` reports Moq1001 on the `(42)` span; `Mock<Foo> f = new(1, true);` reports Moq1002 on `(1, true)`; valid implicit-new rows stay silent.
- [ ] The explicit-form pre-filter (`GetGenericNameSyntax`) still short-circuits non-generic explicit creations.
- [ ] All existing rows in every `ConstructorArgumentsShouldMatchAnalyzerTests.*` partial pass unchanged (both Moq versions, both namespaces).
- [ ] `docs/rules/Moq1001.md` and `docs/rules/Moq1002.md` include a target-typed `new` example.
- [ ] Build 0 warnings including `/p:PedanticMode=true`; `dotnet format` clean.

## Out of scope

- Migrating this analyzer to `RegisterOperationAction(OperationKind.ObjectCreation)`/`OperationKind.Invocation` (ADR-007; also audit items L4/A-13) — a larger refactor that would additionally reduce per-node binding; the syntax-kind addition here is deliberately minimal and does not block that migration.
- The params-collection crash guard in the same analyzer (separate issue).
- Detecting user subclasses of `Mock<T>` (`class MyMock : Mock<IFoo>`) — existing, deliberate scope limit of the `Mock1` symbol gate.
- Deferring `FormatArguments()` allocation to the report path (audit item A-13).
